### PR TITLE
Add a section about CORS

### PR DESCRIPTION
**Pull request by timostamm**
Closes https://github.com/connectrpc/connectrpc.com/issues/42.

### CODE REVIEW

File: docs/cors.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+If the first requirement is not met, a preflight request is issued. If the second
+requirement is not met, the request will error in the browser.
+
+
+## Avoid wildcards
+
+While many CORS headers allow a wildcard `*`, they only do for requests without 
+credentials, such as cookies, TLS client certificates, or authorization headers.
+
+We recommend to avoid wildcards, especially in a production environment. CORS
```
Comment (akshayjshah) [inline docs/cors.md:133]:
```suggestion
It's best to avoid wildcards in production. CORS 
```

**Review by akshayjshah — CHANGES_REQUESTED** [review]
This is great! I left some detailed comments, but overall I'd really like more links to MDN. We should link to the overall CORS page early, and we should link directly to the default header safelists.

File: docs/cors.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+### Allowing origins
+
+`Access-Control-Allow-Origin: https://connectrpc.com` allows browsers to make 
+requests from connectrpc.com via https with the default port 443. 
+
+
+### Allowing methods and headers
+
+`Access-Control-Allow-Methods: POST, GET` is required to support gRPC-web and
```
[the remark this answers is not in the document]
Comment (timostamm) [reply docs/cors.md]: Updated in https://github.com/connectrpc/connectrpc.com/pull/93/commits/a24a357c22daddc17a2a025d126123fdfc0dcbf3

File: docs/cors.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+requests from connectrpc.com via https with the default port 443. 
+
+
+### Allowing methods and headers
+
+`Access-Control-Allow-Methods: POST, GET` is required to support gRPC-web and 
+Connect. gRPC-web will only use POST. Connect uses POST and optionally GET.
+
+`Access-Control-Allow-Headers: Content-Type,Connect-Protocol-Version,Connect-Timeout-Ms,Grpc-Timeout,X-Grpc-Web,X-User-Agent`
+allows all necessary headers for Connect and gRPC-web.
```
[the remark this answers is not in the document]
Comment (timostamm) [reply docs/cors.md]: Updated in https://github.com/connectrpc/connectrpc.com/pull/93/commits/18711745ec65a41dbba37db8922a0922ae82491c

File: docs/cors.md [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+> **Note:** If your application uses custom request headers, you must explicitly 
+> allow them. This includes commonly used headers such as `Authorization`.
+
+
+### Caching preflight responses
+
+`Access-Control-Max-Age: 7200` lets browsers cache CORS information for longer,
```
[the remark this answers is not in the document]
Comment (timostamm) [reply docs/cors.md:79]: Explained in https://github.com/connectrpc/connectrpc.com/pull/93/commits/fb805135d536c618ab34a2644c446774d3ab8b49

File: docs/cors.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+`Access-Control-Max-Age: 7200` lets browsers cache CORS information for longer, 
+which reduces the number of preflight requests. Any changes to your CORS setup 
+won't take effect until the cached data expires. Firefox caps this value at 24h, 
+and modern Chrome caps it at 2h.
+
+`Vary: Origin,Access-Control-Request-Method,Access-Control-Request-Headers` is 
+used by the browser to cache CORS information based on the given header names.
+To avoid poisoning intermediate caches, all request headers resulting in
+conditional preflight responses must be added to the `Vary` response header.
```
[the remark this answers is not in the document]
Comment (timostamm) [reply docs/cors.md]: Updated in https://github.com/connectrpc/connectrpc.com/pull/93/commits/33c2d3db885d7592445608c091a2d7bece769f22

File: docs/cors.md [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+Here is an example for a CORS preflight response for the preflight above that
+allows all requests with the Connect and gRPC-web protocols, and asks the browser
+to cache the information:
+
+```
+HTTP/1.1 204 No Content
+Access-Control-Allow-Origin: https://connectrpc.com
+Access-Control-Allow-Methods: POST, GET
+Access-Control-Allow-Headers: Content-Type,Connect-Protocol-Version,Connect-Timeout-Ms,Grpc-Timeout,X-Grpc-Web,X-User-Agent
```
[the remark this answers is not in the document]
Comment (timostamm) [reply docs/cors.md:51]: Was not deliberate. Removed the space here for consistency.

**Comment by timostamm** [comment]
Thanks for the comments! Updated.

> We should link to the overall CORS page early, 

Certainly a good idea 😆.

> and we should link directly to the default header safelists.

Linking to [safelisted response headers](https://developer.mozilla.org/en-US/docs/Glossary/CORS-safelisted_response_header) in https://github.com/connectrpc/connectrpc.com/pull/93/commits/449c4770291616faf85f4cfacef881f2f83a0be4.

Will it actually be helpful to link to to [safelisted request headers](https://developer.mozilla.org/en-US/docs/Glossary/CORS-safelisted_request_header)? Content-Type is listed there, but it's quite easy to miss the extremely limiting restrictions.

**Comment by akshayjshah** [comment]
> > and we should link directly to the default header safelists.
> 
> Linking to [safelisted response headers](https://developer.mozilla.org/en-US/docs/Glossary/CORS-safelisted_response_header) in [449c477](https://github.com/connectrpc/connectrpc.com/commit/449c4770291616faf85f4cfacef881f2f83a0be4).
> 
> Will it actually be helpful to link to to [safelisted request headers](https://developer.mozilla.org/en-US/docs/Glossary/CORS-safelisted_request_header)? Content-Type is listed there, but it's quite easy to miss the extremely limiting restrictions.

Fair enough. I'm happy with the links as you've proposed them :)

**Review by akshayjshah — APPROVED** [review]

**Pull request merged**